### ISSUE TITLE
JsonReaderException during JSON deserialization

### ISSUE THREAD

**Issue by mu88** [dotnet/runtime#62427]
### Description

From time to time, I'm encountering the following exception when deserialising on of our JSON files:
```
`---- System.FormatException : Could not parse the JSON file.
-------- System.Text.Json.JsonReaderException : Expected end of string, but instead reached end of data. LineNumber: 112 | BytePositionInLine: 13.
   at Microsoft.Extensions.Configuration.FileConfigurationProvider.HandleException(ExceptionDispatchInfo info)
   at Microsoft.Extensions.Configuration.FileConfigurationProvider.Load(Boolean reload)
   at Microsoft.Extensions.Configuration.FileConfigurationProvider.Load()
   at Microsoft.Extensions.Configuration.ConfigurationRoot..ctor(IList`1 providers)
   at Microsoft.Extensions.Configuration.ConfigurationBuilder.Build()
   at Microsoft.Extensions.Hosting.HostBuilder.BuildAppConfiguration()
   at Microsoft.Extensions.Hosting.HostBuilder.Build()
   at <<Somewhere within our code>>) in <<<Directory on our CI server>>
----- Inner Stack Trace -----
   at Microsoft.Extensions.Configuration.Json.JsonConfigurationProvider.Load(Stream stream)
   at Microsoft.Extensions.Configuration.FileConfigurationProvider.Load(Boolean reload)
----- Inner Stack Trace -----
   at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
   at System.Text.Json.Utf8JsonReader.ConsumeString()
   at System.Text.Json.Utf8JsonReader.ConsumePropertyName()
   at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
   at System.Text.Json.Utf8JsonReader.Read()
   at System.Text.Json.JsonDocument.Parse(ReadOnlySpan`1 utf8JsonSpan, JsonReaderOptions readerOptions, MetadataDb& database, StackRowStack& stack)
   at System.Text.Json.JsonDocument.Parse(ReadOnlyMemory`1 utf8Json, JsonReaderOptions readerOptions, Byte[] extraRentedArrayPoolBytes, PooledByteBufferWriter extraPooledByteBufferWriter)
   at System.Text.Json.JsonDocument.Parse(ReadOnlyMemory`1 json, JsonDocumentOptions options)
   at System.Text.Json.JsonDocument.Parse(String json, JsonDocumentOptions options)
   at Microsoft.Extensions.Configuration.Json.JsonConfigurationFileParser.ParseStream(Stream input)
   at Microsoft.Extensions.Configuration.Json.JsonConfigurationFileParser.Parse(Stream input)
   at Microsoft.Extensions.Configuration.Json.JsonConfigurationProvider.Load(Stream stream)
```

The problem mostly occurs on our CI server (1 of 10 times), but I can also reproduce it sometimes on my dev machine when executing the corresponding unit test 500 times.

### Reproduction Steps

Run the following code 500 times:
```c#
new HostBuilder()
    .ConfigureAppConfiguration(config => 
        config.AddJsonFiles(<<Our JSON directory>>))
    .ConfigureWebHostDefaults(
        webHostBuilder =>
        {
            webHostBuilder
                .UseSerilog()
                .UseStartup<Startup>();
        })
    .Build();
```

### Expected behavior

The JSON deserialization works at all times.

### Actual behavior

The JSON deserialization fails locally in 1 of 500 times, on our CI environment in 1 of 10 times.

### Regression?

I recently switched over from Json.NET to `System.Text.Json`

### Known Workarounds

```c#
private async Task<IHost> BuildHostSafely()
{
    var stopwatch = Stopwatch.StartNew();
    while (stopwatch.Elapsed.TotalMilliseconds <= 500)
    {
        try
        {
            return new HostBuilder()
                    .ConfigureAppConfiguration(config => 
                        config.AddJsonFiles(<<Our JSON directory>>))
                    .ConfigureWebHostDefaults(
                        webHostBuilder =>
                        {
                            webHostBuilder
                                .UseSerilog()
                                .UseStartup<Startup>();
                        })
                    .Build();
        }
        catch (InvalidDataException e) when (e.InnerException is FormatException)
        {
            await Task.Delay(50);
        }
    }

    throw new Exception("Mondays are evil");
}
```

### Configuration

- .NET 6 SDK v6.0.100
- Windows 10 Enterprise 20H2 (Build 19042)
- x64

### Other information

I don't want to post the JSON file here, but I can send it to you via mail.

**Comment by eiriktsarpalis** [comment]
The error message seems to suggest that the application is reading partial JSON content. Unfortunately there's not much that can be done on our side unless you can provide us with a deterministic reproduction of the issue. I would suggest making an attempt to log the JSON that is being read, for example by passing a custom implementation of `FileConfigurationProvider` that logs the JSON payload on exception.

**Comment by eerhardt** [comment]
Is something changing/deleting the file? If the file is a well-formed JSON file that isn't changing, I'm not sure how this can happen. Here's where we are reading the file (assuming it is reading a physical file):

https://github.com/dotnet/runtime/blob/12a8819eee9865eb38bca6c05fdece1053102854/src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs#L79-L88

That stream then gets passed directly into this code:

https://github.com/dotnet/runtime/blob/12a8819eee9865eb38bca6c05fdece1053102854/src/libraries/Microsoft.Extensions.Configuration.Json/src/JsonConfigurationFileParser.cs#L22-L33

We are calling `reader.ReadToEnd()`, which reads the whole file into a `string`, and then that `string` is passed to `JsonDocument.Parse` which is throwing the exception: 

`System.Text.Json.JsonReaderException : Expected end of string, but instead reached end of data. LineNumber: 112 | BytePositionInLine: 13.`

What data is at that line / byte position?

**Comment by mu88** [comment]
Thanks for your feedback!

So the JSON file contains both comments and placeholders for our Octopus deployment and looks similar to this:
```json
{
  "SalesProcess": {
    "DaysForAccountBalanceUpdate": #{DaysForAccountBalanceUpdate},
    "TestEmailReceiver": "#{TestEmailReceiver}",
    "SalesReport": {
      "FileTypeKeyMonthlyReport": "MONTHLYREPORTS",
      "DefaultValues": {
        "EmailSender": "#{SalesReportEmailSender}",
        "EmailSenderPFSC": "test@outlook.com",
        /* some 
            multiline comments */
        "EmailSender": "#{SalesReportEmailSender}",
      }
    }
  }
}
```

I've set `JsonSerializerOptions.ReadCommentHandling` to `JsonCommentHandling.Skip` - is `LineNumer: 112` the line number _with_ or _without_ the comments? And is the byte position equivalent to the char position?

**Comment by eerhardt** [comment]
> I've set JsonSerializerOptions.ReadCommentHandling to JsonCommentHandling.Skip - is LineNumer: 112 the line number with or without the comments? 

I believe the line number and byte position is including the comments. @eiriktsarpalis @layomia - please correct me if I'm wrong.

> And is the byte position equivalent to the char position?

In this case, all ASCII characters, you can assume they are the same. In UTF-8, all ASCII characters only take up 1 byte each.

> but I can also reproduce it sometimes on my dev machine when executing the corresponding unit test 500 times.

If you can repro this on your dev machine, can you attach a debugger when it happens and see what `string` is being passed into `JsonDocument.Parse(reader.ReadToEnd(), jsonDocumentOptions))` on line 31 above? That will help figure out if this is a parsing problem, or a file reading problem.

**Comment by mu88** [comment]
It drives me crazy: when debugging the code, I cannot reproduce it, only during run. So I placed a `Debugger.Launch()` statement in the `catch` block. That works, but I'm only getting the exception containing the call stack as written in my very first message. Of course, this is after the fact and I don`t know the `string` content.

Any ideas how I can plug into the deserialization process so that I can answer your question?

**Comment by eerhardt** [comment]
One thought is the approach @eiriktsarpalis outlines above:

https://github.com/dotnet/runtime/issues/62427#issuecomment-986882848

> I would suggest making an attempt to log the JSON that is being read, for example by passing a custom implementation of FileConfigurationProvider that logs the JSON payload on exception.

Basically, copy the code from https://github.com/dotnet/runtime/tree/main/src/libraries/Microsoft.Extensions.Configuration.Json/src into your project, and tweak line 31 of JsonConfigurationFileParser.cs to instead be:

```C#
var contents = reader.ReadToEnd();
try
{
    using (JsonDocument doc = JsonDocument.Parse(contents, jsonDocumentOptions)) 
....
}
catch
{
   // write contents somewhere - console/file/etc
}
```

**Comment by mu88** [comment]
Thanks @eerhardt , I followed your approached and tweaked the code like this:
```c#
var content = string.Empty;
try
{
    content = reader.ReadToEnd();
    using (JsonDocument doc = JsonDocument.Parse(content, jsonDocumentOptions))
    {
        if (doc.RootElement.ValueKind != JsonValueKind.Object)
        {
            throw new FormatException("Error_InvalidTopLevelJSONElement");
        }
        VisitElement(doc.RootElement);
    }
}
catch (Exception e)
{
    var fileContent = $"Exception Type: {e.GetType()}{System.Environment.NewLine}" +
                      $"Exception Message: {e.Message}{System.Environment.NewLine}"+
                      $"Inner Exception Type: {e.InnerException?.GetType()}{System.Environment.NewLine}"+
                      $"Inner Exception Message: {e.InnerException?.Message}{System.Environment.NewLine}" +
                      $"Content:{System.Environment.NewLine}" +
                      $"{content}";
    File.WriteAllText($@"C:\temp\Parse_{DateTime.Now:yyyy-M-dd_HH-mm-ss}.log", fileContent);

    throw;
}
```

I encountered the following error: `Expected depth to be zero at the end of the JSON payload. There is an open JSON object or array that should be closed.`  
And the JSON parser is right: when looking at the loaded content, I see a partially loaded JSON object - many lines from the end of the JSON file are missing.

So it seems as `var contents = reader.ReadToEnd();` does not read the whole JSON file - sometimes 🙄

I played around with the content of the JSON file and it seems to stabilize when shortening the file by removing JSON nodes - but I don't see any pattern yet.  
However it gets more and more reproducable: say I remove one JSON node, then the line number in the exception is 120. I remove another JSON node, line number is e. g. 113. I re-add the previously removed node, line number is 120 again.

I've attached a sample output file. I had to replace some sensitive data, but I did it character by charater - the number of characters is the same as in my production file:  
[Parse_2021-12-09_15-23-27.log](https://github.com/dotnet/runtime/files/7685793/Parse_2021-12-09_15-23-27.log)

**Comment by eiriktsarpalis** [comment]
> So it seems as var contents = reader.ReadToEnd(); does not read the whole JSON file - sometimes 🙄

A more probable explanation is that something is writing to those JSON files concurrently resulting in partial documents being loaded.

**Comment by mu88** [comment]
But how likely is it that two different processes always write to the file at exactly the same time, so that the same line is affected? That seems rather unlikely to me

**Comment by eerhardt** [comment]
Is something modifying the file at all? If yes - who and why?

Another thing you could try is to copy out the base `FileConfigurationProvider`, hook it up to your Json code above and change the following:

https://github.com/dotnet/runtime/blob/8ae1c4972610722013e6664fe59403a176a8d81a/src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs#L78-L89

To not allow concurrent writes to the file. That way you know someone isn't writing to the file at the same time.

**Comment by mu88** [comment]
I was able to resolve the issue 🥳 There were no concurrent writes to the file - I even set the `ReadOnly` flag to the file. When doing so, I discovered the following code which is responsible for creating the file that is being read by `System.Text.Json`: `File.WriteAllTextAsync()` - no `await` 🤦🏻‍♂️

After changing this to `await File.WriteAllTextAsync`, it works like a charm.

Thank you so much for your help!
